bigtable/client/row_reader.h

```diff
@@ -0,0 +1,173 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_ROW_READER_H_
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_ROW_READER_H_
+
+#include "bigtable/client/version.h"
+
+#include "bigtable/client/data_client.h"
+#include "bigtable/client/filters.h"
+#include "bigtable/client/internal/readrowsparser.h"
+#include "bigtable/client/row.h"
+#include "bigtable/client/row_set.h"
+#include "bigtable/client/rpc_backoff_policy.h"
+#include "bigtable/client/rpc_retry_policy.h"
+
+#include <google/bigtable/v2/bigtable.grpc.pb.h>
+
+#include <absl/memory/memory.h>
+#include <absl/types/optional.h>
+#include <grpc++/grpc++.h>
+
+#include <iterator>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+/**
+ * Object returned by Table::ReadRows(), enumerates rows in the response.
+ *
+ * Iterate over the results of ReadRows() using the STL idioms.
+ */
+class RowReader {
+  class RowReaderIterator;
+
+ public:
+  /// Signifies that there is no limit on the number of rows to read.
+  static int constexpr NO_ROWS_LIMIT = 0;
+
+  RowReader(std::shared_ptr<DataClient> client, absl::string_view table_name,
+            RowSet row_set, int rows_limit, Filter filter,
+            std::unique_ptr<RPCRetryPolicy> retry_policy,
+            std::unique_ptr<RPCBackoffPolicy> backoff_policy);
+
+  using iterator = RowReaderIterator;
+
+  /**
+   * Input iterator over rows in the response.
+   *
+   * The returned iterator is a single-pass input iterator that reads
+   * rows from the RowReader when incremented. The first row may be
+   * read when the iterator is constructed.
+   *
+   * Creating, and particularly incrementing, multiple iterators on
+   * the same RowReader is unsupported and can produce incorrect
+   * results.
+   *
+   * Retry and backoff policies are honored.
+   *
+   * @throws std::runtime_error if the read failed after retries.
+   */
+  iterator begin();
+
+  /// End iterator over the rows in the response.
+  iterator end();
+
+  /**
+   * Stop the read call and clean up the connection.
+   *
+   * Invalidates iterators.
+   */
+  void Cancel();
+
+ private:
+  /**
+   * Read and parse the next row in the response.
+   *
+   * @param row receives the next row on success, and is reset on failure or if
+   * there are no more rows.
+   *
+   * This call possibly blocks waiting for data until a full row is available.
+   */
+  void Advance(absl::optional<Row>& row);
```

Comment:
It's very similar to AdvanceOrFail, except that one also returns a Status. Do we have some StatusOr type? I wouldn't like to reinvent it. Overall, I'd personally prefer keeping the out argument as it is, it avoids worrying about some assignments and keeps the code a few lines shorter.

Do we have a way to document the "out" parameters? I believe the GSG mandates pointers, but our style guide asks for references wherever possible, so that kind of read hint is lost. Should we go for an `/* out */` inline comment? Or use pointers?